rococo/repositories/mongodb/mongodb_repository.py

```diff
@@ -0,0 +1,147 @@
+                    field_value = str(field_value.entity_id).replace('-', '')
+                elif isinstance(field_value, dict):
+                    field_value = str(field_value.get('entity_id')).replace('-', '')
+                elif isinstance(field_value, str):
+                    field_value = field_value.replace('-', '')
+            if isinstance(field_value, UUID):
+                field_value = str(field_value).replace('-', '')
+            if isinstance(field_value, datetime):
+                field_value = field_value.strftime('%Y-%m-%d %H:%M:%S')
+
+            data[field.name] = field_value
+        return data
+
+    def save(self, instance: VersionedModel, send_message: bool = False):
+        """Save func"""
+        data = self._process_data_before_save(instance)
+        with self.adapter:
+            inst = self.get_one(self.table_name, "entity_id", {'entity_id': data.get('entity_id')})
+            self.db[f"{self.table_name}_audit"].insert_one(inst)
+            self.db[f"{self.table_name}"].get_one_and_update(
+                {'entity_id': data.get("entity_id")},
+                {'$set': data},
+            )
+            if send_message:
+                # This assumes that the instance is now in post-saved state with all the new DB updates
+                message = json.dumps(instance.as_dict(convert_datetime_to_iso_string=True))
+                self.message_adapter.send_message(self.queue_name, message)
+        return instance
+
+    def _insert(self, data: Dict, collection_name: str):
+        self.db[collection_name].insert_one(data)
+
+    def _outdate(self, data: Dict, collection_name: str):
+        collection = self.db[collection_name]
+        return collection.find_one_and_update(
+            {'entity_id': data.get("entity_id"), 'latest': True, 'active': True},
+            {'$set': {'latest': False}},
+            upsert=False,
+            return_document=ReturnDocument.AFTER
+        )
+
+    def create(self, data: Dict, collection_name: str):
+        return self.update(data, collection_name)
+
+    def create_many(self, data: List[Dict], collection_name: str):
+        self.db[collection_name].insert_many(documents=data)
+
+    def update(self, data: Dict, collection_name: str):
+        with self.client.start_session() as session:
+            with session.start_transaction():
+                try:
+                    updated_doc = self._outdate(data, collection_name)
+                    if updated_doc:
+                        data.update({
+                            'previous_version': updated_doc.get('version'),
+                            'version': get_uuid4_hex(),
+                            'changed_on': datetime.utcnow().isoformat()
+                        })
+                    self._insert(data, collection_name)
+                    return data
+                except Exception as ex:
+                    session.abort_transaction()
```

Comment:
@anikiforov @koteakhaladze We should never retrieve data in the process of a save - all the data necessary for saving comes from the incoming data.  The previous_version is data.version since that represents the version the user was working with when they made their changes.  If we retrieve the "latest" version instead of using the actual version that was sent by the user's action, we have broken our audit trail.